diff -Naur ../Theseus/applications/hello_world/Cargo.toml lab2/Theseus/applications/hello_world/Cargo.toml
--- ../Theseus/applications/hello_world/Cargo.toml	1969-12-31 16:00:00.000000000 -0800
+++ lab2/Theseus/applications/hello_world/Cargo.toml	2021-11-15 19:03:09.148664752 -0800
@@ -0,0 +1,20 @@
+[package]
+name = "hello_world"
+version = "0.1.0"
+edition = "2018"
+authors = ["Ke Wang <ke.wang.kw754@yale.edu>"]
+build = "../../build.rs"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+
+[dependencies.log]
+version = "0.4.8"
+
+
+[dependencies.terminal_print]
+path = "../../kernel/terminal_print"
+
+# [dependencies.application_main_fn]
+# path = "../../compiler_plugins"
\ No newline at end of file
diff -Naur ../Theseus/applications/hello_world/src/lib.rs lab2/Theseus/applications/hello_world/src/lib.rs
--- ../Theseus/applications/hello_world/src/lib.rs	1969-12-31 16:00:00.000000000 -0800
+++ lab2/Theseus/applications/hello_world/src/lib.rs	2021-11-15 19:03:09.152664872 -0800
@@ -0,0 +1,12 @@
+#![no_std]
+// #![feature(plugin)]
+// #![plugin(application_main_fn)]
+
+// #[macro_use] extern crate log;
+#[macro_use] extern crate terminal_print;
+
+pub fn main() -> isize {
+    println!("Hello, world!");
+
+    0
+}
\ No newline at end of file
